New HTML parser

pavpanchekha opened the pull request:
This closes most of the issues we had with the HTML parser, and reorganizes the entire thing.

chrishtr commented on book/html.md:
````diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
 
-``` {.python}
-class ElementNode:
+This special tag is called a [doctype][html5-doctype] that's s always the very first thing in an HTML document. But it's not really an element at all, nor is it supposed to have close tag. Our toy browser won't be using the doctype for anything, so it's best to throw it away:[^quirks-mode]
````
```suggestion
This special tag is called a [doctype][html5-doctype] that's always the very first thing in an HTML document. But it's not really an element at all, nor is it supposed to have close tag. Our toy browser won't be using the doctype for anything, so it's best to throw it away:[^quirks-mode]
```

chrishtr commented on book/html.md:
````diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
- <link rel="next" href="layout" />,
- <link rel="stylesheet" href="../book.css" />]
-```
+Just throwing out doctypes isn't quite enough though---if you run your parser now, it will crash. That's because after the doctype comes a newline, which our parser treats as text and tries to insert into the tree. Except there isn't a tree, since there haven't been any open tags. For simplicity, let's just have our browser skip whitespace-only text nodes to side-step the problem:[^ignore-them]
````
```suggestion
Just throwing out doctypes isn't quite enough though--if you run your parser now, it will crash. That's because after the doctype comes a newline, which our parser treats as text and tries to insert into the tree. Except there isn't a tree, since there haven't been any open tags. For simplicity, let's just have our browser skip whitespace-only text nodes to side-step the problem:[^ignore-them]
```

pavpanchekha replied:
This one is correct, actually. `pandoc` converts triple-dash to an em dashes (for pauses in sentences) and double-dash to en dashes (for ranges).

chrishtr commented on src/lab4.py:
```diff
@@ -54,124 +54,125 @@ def request(url):
-    def __init__(self, text):
+    def __repr__(self):
+        attrs = [k + "=\"" + v + "\"" for k, v  in self.attributes.items()]
+        return "<" + self.tag + " " + " ".join(attrs) + ">"
```
Optimize to avoid the extra space before the > or between / and >; looks ugly in the printed output and is not what real browsers or xml printers ever show.

chrishtr commented on book/html.md:
````diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
+Run it on this web page, and you'll see something like this:
+
+``` {.example}
+ <!DOCTYPE html >
````
Change code to get rid of these extra spaces before the >; left another comment in the code file for this as well.

chrishtr commented on book/html.md:
```diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
+=================
+
+Elements like `<meta>` and `<link>` are what are called self-closing: you don't
+ever write `</meta>` or `</link>`, because these tags don't surround content. To get a reasonable-looking tree out of our parser, we'll need special support for them. In HTML, there's a [fixed list][html-void-elements] of these self-closing tags:[^void-elements]
```
```suggestion
ever write `</meta>` or `</link>`, because these tags don't surround content. To get a reasonable-looking tree out of our parser, we'll need special support for them. In HTML, there's a [fixed list][html5-void-elements] of these self-closing tags:[^void-elements]
```

chrishtr commented on book/html.md:
```diff
@@ -474,101 +584,95 @@ browsers insert them automatically. To do so, they compare the current
 token's tag to the list of currently open elements; that reveals
 whether any additional elements need to be created:
```
```suggestion
whether any additional elements need to be created.
```

pavpanchekha commented:
I think I addressed all of your comments, @chrishtr. Here's a few things it would be good to have your comment on before calling this done:

- Right now `print_tree` is not actually a pretty-printer—it adds quotes and doesn't print close tags. I should rewrite it to actually pretty print, right?
- Is the syntax highlighting exercise still doable? I like it quite a bit, but the lexer is now bound very tightly to the parser, so I'm not sure it's still doable.
- I'm going to do some automated testing on the parser using your Chrome debugging tools branch before committing. There might be easy ways to improve compliance.

chrishtr commented:
If the comments are addressed then LGTM!

Pretty print: Yes it should pretty-print and match the output from the node script.

Syntax highlighting: why is it not still doable exactly? I think it should be possible with a little more refactoring or some choice lines of code?

pavpanchekha commented:
It's doable but I think you'll need to copy-and-paste the lexer. And if you want to highlight attribute values separately then you'll need to merge it with `get_attributes`. Instead of reusing the lexer directly. That might be fine—I do like the syntax highlighting exercise.

chrishtr commented:
I think it's ok; let's put a task for us to do the exercises ourselves and verify how hard it is. We could make some small changes to the chapter code to make it a bit easier.

pavpanchekha commented:
Ok, I've done several passes over the text and it looks good. I've also tested against the Chrome parser and it works well enough on well behaved pages—that's probably all we can promise. I'm ready to ship.

chrishtr commented on book/html.md:
````diff
@@ -8,205 +8,378 @@ next: layout
+``` {.example}
+ <!DOCTYPE HTML>
+   '\n'
+   <HTML LANG="EN-US" XML:LANG="EN-US">
````
All uppercase? sigh.

chrishtr commented on book/html.md:
```diff
@@ -8,205 +8,378 @@ next: layout
+         <META NAME="VIEWPORT" CONTENT="WIDTH=DEVICE-WIDTH,INITIAL-SCALE=1.0,USER-SCALABLE=YES" />
+           <META NAME="AUTHOR" CONTENT="PAVEL PANCHEKHA &AMP; CHRIS HARRELSON" />
+             <LINK REL="STYLESHEET" HREF="BOOK.CSS" />
+               <LINK REL="STYLESHEET" HREF="HTTPS://FONTS.GOOGLEAPIS.COM/CSS?FAMILY=VOLLKORN%7CLORA&DISPLAY=SWAP" />
```
URLs are case-sensitive. I think you will need to remove the upppercase code in the parser.

chrishtr commented on book/html.md:
```diff
@@ -296,168 +472,140 @@ if "=" in attrpair:
-error, but one of the nicer time-saving innovations is *implicit* tags.
+The parser now handles HTML pages correctly—at least when the HTML is
+written by the sorts of goody-two-shoes programmers who remember the
+`<head>` tag, close every open tag, and make their bed. Mere mortals
```
```suggestion
`<head>` tag, close every open tag, and make their bed in the morning. Mere mortals
```

pavpanchekha commented:
The upper-case code was in the pretty-printer, not the parser, luckily. I added it so I could test against the headless Chrome, but in retrospect it's better for the case-conversion code to live in that script, not in the book's code.

pavpanchekha commented:
Please merge if you think this is ready!
